### ISSUE TITLE
Error Handling when pdu is being updated.

### ISSUE THREAD

**Issue by ghost** [psyinfra/prometheus-raritan-pdu-exporter#3]
I was happily testing away on this exporter when I updated the software on one. When the pdu is down this is the error from stderr. 

```
[2023-06-06 09:57:44,709] ERROR: (pdu-idfel-1-red1#-) Uncaught Exception: 0, message='Attempt to decode JSON with unexpected mimetype: text/html', url=URL('https://pdu-idfel-1-red1.admin.evergreen.laika.com/bulk')
```
The exporter itself just acts like the pdu doesn't exist.

If you are monitoring you probably want to know when something is unreachable and you definitely want something more useful than "I couldn't parse the json".

**Comment by ghost** [comment]
When upgrading from 3.4 to 3.5 this is better.
```
ddavis@fruit experiments % /usr/local/bin/prometheus_raritan_pdu_exporter -c prometheus-raritan-pdu-exporter/config.json&[2023-06-06 11:08:13,750] ERROR: (pdu-idfel-1-blue1#-) Uncaught Exception: Cannot connect to host pdu-idfel-1-blue1.admin.evergreen.laika.com:443 ssl:default [Connect call failed ('10.4.98.195', 443)]
```
It may be related to the newer firmware or the states of limbo that the pdus go through as they are updating.

**Comment by ghost** [comment]
This is minor. Ignore it.

**Comment by nhjjr** [comment]
An "uncaught exception" is logged when an error happens we're not specifically looking out for. In this case, a readout of the PDU via its API has failed. Importantly, the PDU _is_ reachable, but it is not providing the expected JSON output when its API is queried. Had it not been reachable a different warning would have been logged. I expect the same as you that this is likely caused by the PDU going through its update process (or any other transitory state) and should ultimately self-correct. I think the error logged here is as helpful as it can be.

Note that if this were to happen during the startup process of the exporter it will exit with an error, because that process is necessary to discover the PDU's connectors.

**Comment by ghost** [comment]
Thank you for your quick response and I agree since the issue did go away once the update was complete. 
As for the " Note that if this were to happen during the startup process of the exporter it will exit with an error, because that process is necessary to discover the PDU's connectors." Is there a way to get it to ignore errors and start up if one of the pdu's is missing?

(in the case below the rack is being rewired so we are missing 2 out of 100+)

ddavis@fruit raritan % /usr/local/bin/prometheus_raritan_pdu_exporter -c config.json
[2023-06-07 11:33:16,333] ERROR:
[2023-06-07 11:33:16,333] CRITICAL: Exporter shut down due to an error during the setup procedure. Please contact your administrator.

If there was a way to flag it as down and monitor the rest that would be awesome.

**Comment by nhjjr** [comment]
I made this originally for our own use-case of 6 (now 8) PDUs, so having a PDU unreachable at start-up was unlikely (and if it were the case, temporarily removing it from the config.json would not be difficult). I can imagine that with 100+ PDUs this is not desirable. I can change the startup failing to connect to a PDU to log a warning instead of exit with an error. Changing this behavior to a warning comes with a few caveats though: (a) The missing PDUs will not be scraped until the exporter is restarted, and (b) any changes to a PDU's configuration that affects metrics requires an exporter restart (this is already the case, e.g., if you change an outlet name or add a peripheral device).

The reason for these caveats is because the setup procedure is very straining. It searches query results from various API endpoints and methods on the PDU to find all its connectors and connector meta-data, so that subsequent readouts can query the connectors directly (in bulk). The upside is that readouts are (relatively) fast allowing the use of the standard 15s scrape interval in Prometheus. The downside is that any changes after the exporter has been started won't be picked up until a restart.

What may be of interest to you if version 3.5 of the PDU firmware is not a hard requirement, is that recently Raritan introduced a native Prometheus exporter in version 4.0.10 GA. Their [documentation](https://cdn1.raritan.com/download/pdu-g2/4.0.20/PDU_G2_UG_A2_4.0.20.pdf) explains how this can be used. With 100+ PDUs this may be a preferable approach, as this exporter was primarily built (and tested) using our setup of 6 (now 8) PDUs. We're currently using the native exporter, but have some trouble with uncharacteristically lengthy scrape durations and are in contact with Raritan support about this.

Please let me know if you'd still prefer the above proposed change.

**Comment by ghost** [comment]
Sorry about the lag. We tried to use the native exporter but were unhappy with the details (in particular it's missing the power state of each port).
We would very much like to use your importer with the proposed change in startup behavior.

**Comment by nhjjr** [comment]
I just pushed a new release (v2.1.5) that takes care of this request. 

When a PDU cannot be reached during start-up, you will be given a warning that it cannot be reached, and subsequently another warning that it has been removed from collection. When the PDU is reachable again you only have to restart the exporter without changing the configuration file.

Let me know if you run into any problems. I'll consider this issue closed for now :)
